core/iwasm/interpreter/wasm_loader.c

```diff
@@ -12083,6 +12083,11 @@ wasm_loader_prepare_bytecode(WASMModule *module, WASMFunction *func,
 
                 read_leb_uint32(p, p_end, type_idx);
 #if WASM_ENABLE_REF_TYPES != 0 || WASM_ENABLE_GC != 0
+#if WASM_ENABLE_WAMR_COMPILER != 0
+                if (*p != 0x00) {
+                    module->is_ref_types_used = true;
+                }
+#endif
```

Comment:
Thanks for reporting the issue, seems that it is invalid to read table_idx with `read_uint8` even when REF_TYPES/GC isn't enabled, from what spec mentions and what other places handle (wasm_interp_classic.c CALL_INDIRECT also read table_idx with read_leb_uint32):
https://webassembly.github.io/spec/core/binary/modules.html#binary-tableidx,
we should always treat table_idx as uint32 and use read_leb_uint32 to read the data, and after that, check whether table_idx is 0: if not, then set `module->is_ref_types_used = true`:
```C
                read_leb_uint32(p, p_end, table_idx);
#if WASM_ENABLE_REF_TYPES != 0 || WASM_ENABLE_GC != 0
#if WASM_ENABLE_WAMR_COMPILER != 0
                if (table_idx != 0)
                    module->is_ref_types_used = true;
#endif
#endif
                if (!check_table_index(module, table_idx, error_buf,
                                       error_buf_size)) {
                    goto fail;
                }
```

And should also change the handlings of CALL_INDIRECT and RETURN_CALL_INDIRECT in many other places: wasm_loader_find_block_addr in this file, wasm_mini_loader.c, aot_compiler.c and fast-jit/jit_frontend.c.

Could you help confirm it and fix the related code accordingly? Thanks.

Reply:
> Thanks for reporting the issue, seems that it is invalid to read table_idx with `read_uint8` even when REF_TYPES/GC isn't enabled, from what spec mentions and what other places handle (wasm_interp_classic.c CALL_INDIRECT also read table_idx with read_leb_uint32): https://webassembly.github.io/spec/core/binary/modules.html#binary-tableidx, we should always treat table_idx as uint32 and use read_leb_uint32 to read the data, and after that, check whether table_idx is 0: if not, then set `module->is_ref_types_used = true`:

our REF_TYPES means https://github.com/WebAssembly/reference-types, right?
the url you are referring to is inappropriate because it already includes reference-types.
looking at the diff, (https://github.com/WebAssembly/spec/commit/7fa2f20a6df4cf1c114582c8cb60f5bfcdbf1be1#diff-4343dbfdb9bd9737d2f391b4624ca72216a55d3ea74eca470aab7fd525ec3958)
the call_indirect imm before reference-types was just 00, not a leb u32 0.

my understanding is that there is no way to distinguish an invalid module (call_indirect .. 0x80 00 w/o reference-types) and a valid redundant encoding. (call_indirect .. 0x80 00 w/ reference-types)
as wamrc has an option to disable reference-types, (`--disable-ref-types`) i guess the best thing we can do is to make wasm_loader.c honor the option when WASM_ENABLE_WAMR_COMPILER is defined.

Reply:
Yes, thanks for the correction, it really easily makes confusing since the spec is evolving. You are right, in the earlier version of wasm spec without reference-types, the table index is 0 of one byte, this is the snapshot from wasm spec Release 1.1 pdf document:
![image](https://github.com/user-attachments/assets/c828eb7a-02fb-425d-bc67-57c628329f24)

Do you mean this PR's fixing of wasm_loader.c is good and we should also handle other places?

Reply:
> Do you mean this PR's fixing of wasm_loader.c is good and we should also handle other places?

i want to make wamrc bail out when enable_ref_types==false and is_ref_types_used==true at least.
but it doesn't need to be a part of this PR.

Reply:
OK, so maybe we can merge this PR first and then submit another PR to fix the wamrc issue and issues in other places.

Reply:
sure.